Feat/add calc withdraw

jcompagni10 opened the pull request:
Add calculate withdrawable amount to QueryUserLimitOrderTranche

This is needed so that CW contracts can check how much they can withdraw from a limit order. Specifically needed for astroport integration.

jcompagni10 committed:
Add CalcWithdrawable to userLimitOrderTranche Query

jcompagni10 committed:
Add CLI option for calc-withdraw

jcompagni10 committed:
Switch to withdrawableShares + small fixes

NeverHappened reviewed (commented):
integration tests failed. please use sdk-47 branch for neutron-dao

NeverHappened commented on x/dex/keeper/grpc_query_limit_order_tranche_user.go:
```diff
@@ -46,14 +47,34 @@ func (k Keeper) LimitOrderTrancheUserAll(
+	if !found {
+		return math.ZeroInt(), status.Error(codes.NotFound, "Tranche not found")
+	}
+	withdrawableShares, _ := tranche.Withdraw(&trancheUser)
```
How about small refactoring in tranche? we can make new function tranche#Withdrawable (or some kind of this), that will only return stuff and do not modify data. And tranche#Withdraw will use this function and just change `t.ReservesTakerDenom = reservesTokenOutDec.Sub(amountOutTokenOut).TruncateInt()`.
That way there is some boundaries between read and write functions.

mmikev commented on proto/neutron/dex/query.proto:
```diff
@@ -133,10 +133,17 @@ message QueryParamsResponse {
 message QueryGetLimitOrderTrancheUserRequest {
   string address = 1;
   string tranche_key = 2;
+  bool calc_withdrawable_shares = 3;
```
i suppose this is not API breaking cause it's just a fiiled addition, but could you confirm pls?

jcompagni10 replied:
Correct. If a value is not supplied it will default to false and the pre-existing behavior will be maintained

mmikev commented on x/dex/client/cli/query_limit_order_tranche_user.go:
```diff
@@ -45,18 +45,23 @@ func CmdListLimitOrderTrancheUser() *cobra.Command {
+		Use:     "show-limit-order-tranche-user [address] [tranche-key] ?(--calc-withdraw)",
 		Short:   "shows a LimitOrderTrancheUser",
-		Example: "show-limit-order-tranche-user TRANCHEKEY123 alice",
+		Example: "show-limit-order-tranche-user alice TRANCHEKEY123",
```
I think it's better to use valid neutron address as an example instead of `alice`

jcompagni10 committed:
Small fixes

- change cli exmaple to have real neutron addr
- create dedicated calcWithdraw method so there is no risk of altering state

jcompagni10 commented:
https://github.com/neutron-org/neutron-tests/actions/runs/8473082006

jcompagni10 committed:
formatting

jcompagni10 committed:
Merge remote-tracking branch 'origin/main' into feat/add-calc-withdraw
